tests/test_physics_determinism.py

```diff
@@ -0,0 +1,133 @@
+
+    cfg_settings[
+        "scene"
+    ] = "data/scene_datasets/habitat-test-scenes/skokloster-castle.glb"
+    # enable the physics simulator: also clears available actions to no-op
+    cfg_settings["enable_physics"] = True
+    cfg_settings["depth_sensor"] = True
+
+    # test loading the physical scene
+    hab_cfg = examples.settings.make_cfg(cfg_settings)
+    with habitat_sim.Simulator(hab_cfg) as sim:
+        sim.reset()
+        obj_mgr = sim.get_object_template_manager()
+
+        assert obj_mgr.get_num_templates() > 0
+
+        # test adding an object to the world
+        # get handle for object 0, used to test
+        obj_handle_list = obj_mgr.get_template_handles("sphere")
+        sphere_object_id = sim.add_object_by_handle(obj_handle_list[0])
+        sphere_position = mn.Vector3(
+            [-0.9517786502838135, 2.167676642537117, 11.343990325927734]
+        )
+
+        sim.set_object_motion_type(
+            habitat_sim.physics.MotionType.DYNAMIC, sphere_object_id
+        )
+
+        sim.set_translation(sphere_position, sphere_object_id)
+
+        obj_handle_list = obj_mgr.get_template_handles("chair")
+        chair_object_id = sim.add_object_by_handle(obj_handle_list[0])
+
+        chair_position = mn.Vector3(
+            [-0.9517786502838135, 1.57676642537117, 11.343990325927734]
+        )
+
+        sim.set_object_motion_type(
+            habitat_sim.physics.MotionType.DYNAMIC, chair_object_id
+        )
+
+        sim.set_translation(chair_position, chair_object_id)
+
+        len(sim.get_existing_object_ids()) == 3
+
+        states = []
+        world_time = sim.get_world_time()
+        step_count = 1
+        states.append({"world_time": world_time, "step_count": 0, "object_states": []})
+        while world_time < 3.0:
+            sim.step_world()
+            world_time = sim.get_world_time()
+            states.append(
+                {
+                    "world_time": world_time,
+                    "step_count": step_count,
+                    "object_states": get_all_object_states(sim),
+                }
+            )
+            step_count += 1
+
+        ground_truth_states = get_ground_truth()
+
+        assert len(states) == len(ground_truth_states)
+
+        for i in range(len(states)):
+            assert states[i]["world_time"] == ground_truth_states[i]["world_time"]
+            assert states[i]["step_count"] == ground_truth_states[i]["step_count"]
+
+            assert states[i].get("object_states") is not None
+
+            if states[i].get("object_states"):
+                object_states = states[i].get("object_states")
+                ground_truth_object_states = ground_truth_states[i].get("object_states")
+                for idx in range(len(object_states)):
+                    assert (
+                        object_states[idx]["object_id"]
+                        == ground_truth_object_states[idx]["object_id"]
```

Comment:
np.isclose is great for this. @Ram81

Reply:
@erikwijmans @Skylion007 should we use the same on JS side too? The JS test is failing from first physics step because of some floating point error in object translation.

Reply:
@ram81 I think JEST has a utility function https://jestjs.io/docs/en/expect#tobeclosetonumber-numdigits , but you can also quickly code up your own if needed.

Reply:
>A tolerance test of 1e-3 or 1e-5 is more realistic and I would still call that deterministic.
Maybe we are just disagreeing on terminology.  I'm used to treating determinism as all-or-nothing.  I'd consider the simulation deterministic only if it produces exact same results. It seems inevitable that any desync, no matter how small, is going to accumulate/multiply over hundreds/thousands of physics steps into a significant divergence. 

There may be some concept of "soft determinism" where we use epsilons and guarantee the max divergence of a single physics step. But I guess there's no guarantee we could make about the size of divergence over hundreds/thousands of steps.

Reply:
>Floating point math is non-associative, so this precludes us (or any dependency and even the compiler) from re-ordering any math anywhere in the entire pipeline, which isn't realistic.
You're suggesting that a compiler change could result in some C++ floating-point expressions evaluating differently (due to the operations getting reordered by the compiler). While I agree that exact determinism is going to be difficult/impossible, I don't know that this is specifically the problem.

C++ has rules for how to evaluate expressions, including ordering. IEEE floating-point standard dictates the results of individual operations. The compiler shouldn't deviate here. (Although I've seen compiler optimization options that let the compiler do approximate floating-point math--obviously that would fail.)

Reply:
> C++ has rules for how to evaluate expressions, including ordering. IEEE floating-point standard dictates the results of individual operations. The compiler shouldn't deviate here. (Although I've seen compiler optimization options that let the compiler do approximate floating-point math--obviously that would fail.)

What I mean by compile re-ordering is more to do with knock-on effects.  i.e. when dealing with something like `a + b + c`, the compiler chooses `(a + b) + c` or `a + (b + c)` based on what is already in the registers, so seemingly unrelated changes may change what persists in registers and thus floating point math.  Also, some machines will have fused floating point instructions.  These accumulate at a higher precision than emitting multiple instructions and will incur less rounding error leading to weirdness -- `a*b + (-a)*b != 0` is the classic example of this.

> There may be some concept of "soft determinism" where we use epsilons and guarantee the max divergence of a single physics step. But I guess there's no guarantee we could make about the size of divergence over hundreds/thousands of steps.

I was thinking we would do epsilons over hundreds/thousands of steps.  These types of discrepancies should be small enough that they won't result in exploding divergence, but also do exist and make expecting the exact same bit pattern impossible.